### DISCUSSION TITLE
Reusable PostgreSQL dev service for different data sources

### DISCUSSION THREAD

**Discussion by DCCSKrezovic** [quarkusio/quarkus#31655] [Community]
Hello!

I have defined two data sources and would like to use devservices to create a test DB for me.

I have explicitly given, for testing purposes,

```
quarkus.datasource.devservices.enabled=true
quarkus.datasource.schemaowner.devservices.enabled=true
```

where the correct options are set, i.e.

```
quarkus.datasource.db-kind=postgresql
...
quarkus.datasource.schemaowner.db-kind=postgresql
...
```

When I do `mvn quarkus:dev`  or `mvn clean install`, the devservices will create two containers, one for default data source, another for named data source.

Is there a way to use the same devservices container for two data sources?

I require "schemaowner" datasource to specify additional set of credentials for liquibase and not run the whole app with users that have ALTER privileges. The bug report that fixes this is already available, but not the topic right now ... https://github.com/quarkusio/quarkus/issues/31214

**Comment by vipul0805** [comment]
Hi @DCCSKrezovic, I am also facing a similar kind of problem where I have two data sources `default and read-only`, and want to reuse the default database container for the named one. Could not find anything regarding this on the `Quarkus` docs.

I wanted to ask if you have figured out a way to do it. Thanks.

**Comment by mapuci** [comment]
docker compose up -d :)

I had to dig up the same answer a few months ago when I had to connect 2 quarkus apps with different db users to same db. Painful story indeed. Here are my insights (I'm not RedHat-ter)

Some devservices are designed to be shared (e.g. kafka, elastic), databases are (probably for a good reason) not.
See [quarkus.kafka.devservices.shared](https://quarkus.io/guides/dev-services#quarkus-kafka-client-config-group-kafka-dev-services-build-time-config_quarkus.kafka.devservices.shared)

Testcontainers(or quarkus devservices under the hood) are primarily a testing library, featuring disposable containers.
A new container will be created each and every time there is no existing running container with IDENTICAL! config.

And quarkus starts db testcontainer with some env vars that create db,user,password. Thats why, if you specify different users with [quarkus.datasource.devservices.username](https://quarkus.io/guides/dev-services#quarkus-datasource-config-group-dev-services-build-time-config_quarkus.datasource.devservices.username) there will be new container spawned for each service.

If you want you can check out testcontainers docs, here you will find some meaningfull insights (about dockerJava command that is executed to start container): https://java.testcontainers.org/features/advanced_options
Using CreateContainerCmdModifier

My advice is: when using 1app=1db devservices are fine, otherwise go with docker-compose.
In tests you can even start it up yourself using https://java.testcontainers.org/modules/docker_compose/ and https://quarkus.io/guides/getting-started-testing#quarkus-test-resource
For dev mode you will have to start it up yourself. And the db users/urls/passwords can be hardcoded to application.yaml for dev and test profiles in that case.
